// @sitecore-cloudsdk/events/dist/cjs/src/server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PACKAGE_INITIALIZER_METHOD_NAME = exports.PACKAGE_NAME = exports.EVENTS_NAMESPACE = exports.PACKAGE_VERSION = exports.pageView = exports.identity = exports.event = void 0;
// © Sitecore Corporation A/S. All rights reserved. Sitecore® is a registered trademark of Sitecore Corporation A/S.
require("./lib/initializer/server/initializer");
var eventServer_1 = require("./lib/events/custom-event/eventServer");
Object.defineProperty(exports, "event", { enumerable: true, get: function () { return eventServer_1.eventServer; } });
var identityServer_1 = require("./lib/events/identity/identityServer");
Object.defineProperty(exports, "identity", { enumerable: true, get: function () { return identityServer_1.identityServer; } });
var page_view_server_1 = require("./lib/events/page-view/page-view-server");
Object.defineProperty(exports, "pageView", { enumerable: true, get: function () { return page_view_server_1.pageViewServer; } });
var consts_1 = require("./lib/consts");
Object.defineProperty(exports, "PACKAGE_VERSION", { enumerable: true, get: function () { return consts_1.PACKAGE_VERSION; } });
Object.defineProperty(exports, "EVENTS_NAMESPACE", { enumerable: true, get: function () { return consts_1.EVENTS_NAMESPACE; } });
Object.defineProperty(exports, "PACKAGE_NAME", { enumerable: true, get: function () { return consts_1.PACKAGE_NAME; } });
Object.defineProperty(exports, "PACKAGE_INITIALIZER_METHOD_NAME", { enumerable: true, get: function () { return consts_1.PACKAGE_INITIALIZER_METHOD_NAME; } });
