// @sitecore-cloudsdk/events/dist/cjs/src/lib/consts.js
"use strict";
// © Sitecore Corporation A/S. All rights reserved. Sitecore® is a registered trademark of Sitecore Corporation A/S.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorMessages = exports.X_CLIENT_SOFTWARE_ID = exports.PACKAGE_INITIALIZER_METHOD_NAME = exports.PACKAGE_NAME = exports.PACKAGE_VERSION = exports.EVENTS_NAMESPACE = void 0;
/* eslint-disable @typescript-eslint/naming-convention */
const package_json_1 = __importDefault(require("../../package.json"));
const consts_1 = require("./events/consts");
exports.EVENTS_NAMESPACE = 'sitecore-cloudsdk:events';
exports.PACKAGE_VERSION = package_json_1.default.version;
exports.PACKAGE_NAME = package_json_1.default.name;
exports.PACKAGE_INITIALIZER_METHOD_NAME = 'addEvents';
/**
 * Returns the name & version of the library in a String.
 */
exports.X_CLIENT_SOFTWARE_ID = `${exports.PACKAGE_NAME} ${exports.PACKAGE_VERSION}`;
/* eslint-disable max-len */
var ErrorMessages;
(function (ErrorMessages) {
    ErrorMessages["IE_0001"] = "[IE-0001] You are trying to run a browser-side function on the server side. On the server side, run the server-side equivalent of the function, available in \"server\" modules.";
    ErrorMessages["IE_0014"] = "[IE-0014] You must first initialize the Cloud SDK and the \"events\" package. First, import \"CloudSDK\" from \"@sitecore-cloudsdk/core/browser\" and import \"@sitecore-cloudsdk/events/browser\". Then, run \"CloudSDK().addEvents().initialize()\".";
    ErrorMessages["IE_0015"] = "[IE-0015] You must first initialize the Cloud SDK and the \"events\" package. First, import \"CloudSDK\" from \"@sitecore-cloudsdk/core/server\" and import \"@sitecore-cloudsdk/events/server\". Then, run \"await CloudSDK().addEvents().initialize()\".";
    ErrorMessages["IV_0002"] = "[IV-0002] Incorrect value for \"dob\". Format the value according to ISO 8601.";
    ErrorMessages["IV_0003"] = "[IV-0003] Incorrect value for \"email\". Set the value to a valid email address.";
    ErrorMessages["IV_0004"] = "[IV-0004] Incorrect value for \"expiryDate\". Format the value according to ISO 8601.";
    ErrorMessages["IV_0005"] = "[IV-0005] \"extensionData\" supports maximum 50 attributes. Reduce the number of attributes.";
    ErrorMessages["MV_0003"] = "[MV-0003] \"identifiers\" is required.";
})(ErrorMessages || (exports.ErrorMessages = ErrorMessages = {}));
/* eslint-enable max-len */
