// @sitecore-cloudsdk/core/dist/cjs/src/lib/consts.js
"use strict";
// © Sitecore Corporation A/S. All rights reserved. Sitecore® is a registered trademark of Sitecore Corporation A/S.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorMessages = exports.CORRELATION_ID_HEADER = exports.SITECORE_EDGE_URL = exports.API_VERSION = exports.DAILY_SECONDS = exports.DEFAULT_COOKIE_EXPIRY_DAYS = exports.COOKIE_NAME_PREFIX = exports.LIBRARY_VERSION = void 0;
/* eslint-disable @typescript-eslint/naming-convention */
const package_json_1 = __importDefault(require("../../package.json"));
exports.LIBRARY_VERSION = package_json_1.default.version;
exports.COOKIE_NAME_PREFIX = 'sc_';
exports.DEFAULT_COOKIE_EXPIRY_DAYS = 730;
exports.DAILY_SECONDS = 86400;
exports.API_VERSION = 'v1.2';
exports.SITECORE_EDGE_URL = 'https://edge-platform.sitecorecloud.io';
exports.CORRELATION_ID_HEADER = 'x-sc-correlation-id';
/* eslint-disable max-len */
var ErrorMessages;
(function (ErrorMessages) {
    ErrorMessages["IE_0001"] = "[IE-0001] You are trying to run a browser-side function on the server side. On the server side, run the server-side equivalent of the function, available in \"server\" modules.";
    ErrorMessages["IE_0003"] = "[IE-0003] Unable to set the \"sc_{SitecoreEdgeContextId}\" cookie because the browser ID could not be retrieved from the server. Make sure to set the correct values for \"sitecoreEdgeContextId\" and \"siteName\". If the issue persists, try again later or use try-catch blocks to handle this error.";
    ErrorMessages["IE_0011"] = "[IE-0011] Unable to set the \"sc_{SitecoreEdgeContextId}_personalize\" cookie because the guest ID could not be retrieved from the server. Make sure to set the correct values for \"sitecoreEdgeContextId\" and \"siteName\". If the issue persists, try again later or use try-catch blocks to handle this error.";
    ErrorMessages["IE_0012"] = "[IE-0012] You must first initialize the Cloud SDK. Import \"CloudSDK\" from \"@sitecore-cloudsdk/core/browser\", then run \"CloudSDK().initialize()\".";
    ErrorMessages["IE_0013"] = "[IE-0013] You must first initialize the Cloud SDK. Import \"CloudSDK\" from \"@sitecore-cloudsdk/core/server\", then run \"await CloudSDK().initialize()\".";
    ErrorMessages["IV_0001"] = "[IV-0001] Incorrect value for \"sitecoreEdgeUrl\". Set the value to a valid URL string.";
    ErrorMessages["MV_0001"] = "[MV-0001] \"sitecoreEdgeContextId\" is required.";
    ErrorMessages["MV_0002"] = "[MV-0002] \"siteName\" is required.";
})(ErrorMessages || (exports.ErrorMessages = ErrorMessages = {}));
/* eslint-enable max-len */
