// @sitecore-cloudsdk/personalize/dist/cjs/src/lib/consts.js
"use strict";
// © Sitecore Corporation A/S. All rights reserved. Sitecore® is a registered trademark of Sitecore Corporation A/S.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorMessages = exports.UTM_PREFIX = exports.PACKAGE_NAME = exports.PACKAGE_VERSION = exports.PERSONALIZE_NAMESPACE = void 0;
/* eslint-disable @typescript-eslint/naming-convention */
const package_json_1 = __importDefault(require("../../package.json"));
exports.PERSONALIZE_NAMESPACE = 'sitecore-cloudsdk:personalize';
exports.PACKAGE_VERSION = package_json_1.default.version;
exports.PACKAGE_NAME = package_json_1.default.name;
exports.UTM_PREFIX = 'utm_';
var ErrorMessages;
(function (ErrorMessages) {
    // eslint-disable-next-line max-len
    ErrorMessages["IE_0001"] = "[IE-0001] You are trying to run a browser-side function on the server side. On the server side, run the server-side equivalent of the function, available in \"server\" modules.";
    // eslint-disable-next-line max-len
    ErrorMessages["IE_0016"] = "[IE-0016] You must first initialize the Cloud SDK and the \"personalize\" package. First, import \"CloudSDK\" from \"@sitecore-cloudsdk/core/browser\" and import \"@sitecore-cloudsdk/personalize/browser\". Then, run \"CloudSDK().addPersonalize().initialize()\".";
    // eslint-disable-next-line max-len
    ErrorMessages["IE_0017"] = "[IE-0017] You must first initialize the Cloud SDK and the \"personalize\" package. First, import \"CloudSDK\" from \"@sitecore-cloudsdk/core/server\", and import \"@sitecore-cloudsdk/personalize/server\". Then, run \"await CloudSDK().addPersonalize().initialize()\".";
    ErrorMessages["MV_0004"] = "[MV-0004] \"friendlyId\" is required.";
})(ErrorMessages || (exports.ErrorMessages = ErrorMessages = {}));
