// @sitecore-cloudsdk/events/dist/cjs/src/lib/events/identity/identity-event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdentityEvent = void 0;
const utils_1 = require("@sitecore-cloudsdk/utils");
const base_event_1 = require("../base-event");
const consts_1 = require("../../consts");
const consts_2 = require("../consts");
class IdentityEvent extends base_event_1.BaseEvent {
    /**
     * A class that extends from {@link BaseEvent} and has all the required functionality to send a VIEW event
     *
     * @param args - Unified object containing the required properties
     */
    constructor(args) {
        const { channel, currency, language, page, extensionData } = args.identityData;
        super({ channel, currency, language, page }, args.id);
        this.extensionData = {};
        this.numberOfExtensionDataProperties = 0;
        this.validateAttributes(args.identityData);
        this.identityData = args.identityData;
        this.sendEvent = args.sendEvent;
        this.settings = args.settings;
        if (extensionData)
            this.extensionData = (0, utils_1.flattenObject)({ object: extensionData });
        this.numberOfExtensionDataProperties = Object.entries(this.extensionData).length;
        if (this.numberOfExtensionDataProperties > consts_2.MAX_EXT_ATTRIBUTES)
            throw new Error(consts_1.ErrorMessages.IV_0005);
    }
    /**
     * Function that validates the identifiers object, email and date attributes for CDN users
     *  * @param identityData - The data to be validated
     */
    validateAttributes(identityData) {
        if (identityData.identifiers.length === 0)
            throw new Error(consts_1.ErrorMessages.MV_0003);
        if (identityData.dob !== undefined && !(0, utils_1.isShortISODateString)(identityData.dob))
            throw new Error(consts_1.ErrorMessages.IV_0002);
        identityData.identifiers.forEach((identifier) => {
            if (identifier.expiryDate && !(0, utils_1.isShortISODateString)(identifier.expiryDate))
                throw new Error(consts_1.ErrorMessages.IV_0004);
        });
        if (identityData.email && !(0, utils_1.isValidEmail)(identityData.email))
            throw new Error(consts_1.ErrorMessages.IV_0003);
    }
    /**
     * A function that maps the identity event input data with the payload sent to the API
     * @returns - The payload object
     */
    mapAttributes() {
        const identityPayload = {
            city: this.identityData.city,
            country: this.identityData.country,
            dob: this.identityData.dob,
            email: this.identityData.email,
            firstname: this.identityData.firstName,
            gender: this.identityData.gender,
            identifiers: this.identityData.identifiers.map((value) => {
                return {
                    // eslint-disable-next-line @typescript-eslint/naming-convention
                    expiry_date: value.expiryDate,
                    id: value.id,
                    provider: value.provider
                };
            }),
            lastname: this.identityData.lastName,
            mobile: this.identityData.mobile,
            phone: this.identityData.phone,
            // eslint-disable-next-line @typescript-eslint/naming-convention
            postal_code: this.identityData.postalCode,
            state: this.identityData.state,
            street: this.identityData.street,
            title: this.identityData.title,
            type: 'IDENTITY'
        };
        if (this.numberOfExtensionDataProperties > 0)
            identityPayload.ext = this.extensionData;
        return identityPayload;
    }
    /**
     * Sends the event to Sitecore EP
     * @returns - A promise that resolves with either the Sitecore EP response object or null
     */
    async send() {
        const baseAttr = this.mapBaseEventPayload();
        const eventAttrs = this.mapAttributes();
        const fetchBody = Object.assign({}, eventAttrs, baseAttr);
        return await this.sendEvent(fetchBody, this.settings);
    }
}
exports.IdentityEvent = IdentityEvent;
