// @sitecore-cloudsdk/utils/dist/cjs/src/index.js
"use strict";
// © Sitecore Corporation A/S. All rights reserved. Sitecore® is a registered trademark of Sitecore Corporation A/S.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.appendScriptWithAttributes = exports.isValidLocation = exports.isValidHttpURL = exports.ErrorMessages = exports.normalizeHeaders = exports.generateV4UUID = exports.omit = exports.isValidEmail = exports.isShortISODateString = exports.flattenObject = exports.cookieExists = exports.getCookieServerSide = exports.getCookie = exports.isHttpResponse = exports.isHttpRequest = exports.isNextJsMiddlewareResponse = exports.isNextJsMiddlewareRequest = exports.getCookieValueClientSide = exports.fetchWithTimeout = exports.createCookieString = void 0;
var create_cookie_string_1 = require("./lib/cookies/create-cookie-string");
Object.defineProperty(exports, "createCookieString", { enumerable: true, get: function () { return create_cookie_string_1.createCookieString; } });
var fetch_with_timeout_1 = require("./lib/fetch-with-timeout");
Object.defineProperty(exports, "fetchWithTimeout", { enumerable: true, get: function () { return fetch_with_timeout_1.fetchWithTimeout; } });
var get_cookie_value_client_side_1 = require("./lib/cookies/get-cookie-value-client-side");
Object.defineProperty(exports, "getCookieValueClientSide", { enumerable: true, get: function () { return get_cookie_value_client_side_1.getCookieValueClientSide; } });
var is_next_js_middleware_request_1 = require("./lib/typeguards/is-next-js-middleware-request");
Object.defineProperty(exports, "isNextJsMiddlewareRequest", { enumerable: true, get: function () { return is_next_js_middleware_request_1.isNextJsMiddlewareRequest; } });
var is_next_js_middleware_response_1 = require("./lib/typeguards/is-next-js-middleware-response");
Object.defineProperty(exports, "isNextJsMiddlewareResponse", { enumerable: true, get: function () { return is_next_js_middleware_response_1.isNextJsMiddlewareResponse; } });
var is_http_request_1 = require("./lib/typeguards/is-http-request");
Object.defineProperty(exports, "isHttpRequest", { enumerable: true, get: function () { return is_http_request_1.isHttpRequest; } });
var is_http_response_1 = require("./lib/typeguards/is-http-response");
Object.defineProperty(exports, "isHttpResponse", { enumerable: true, get: function () { return is_http_response_1.isHttpResponse; } });
var get_cookie_1 = require("./lib/cookies/get-cookie");
Object.defineProperty(exports, "getCookie", { enumerable: true, get: function () { return get_cookie_1.getCookie; } });
var get_cookie_server_side_1 = require("./lib/cookies/get-cookie-server-side");
Object.defineProperty(exports, "getCookieServerSide", { enumerable: true, get: function () { return get_cookie_server_side_1.getCookieServerSide; } });
var cookie_exists_1 = require("./lib/cookies/cookie-exists");
Object.defineProperty(exports, "cookieExists", { enumerable: true, get: function () { return cookie_exists_1.cookieExists; } });
var flatten_object_1 = require("./lib/converters/flatten-object");
Object.defineProperty(exports, "flattenObject", { enumerable: true, get: function () { return flatten_object_1.flattenObject; } });
var is_short_iso_date_string_1 = require("./lib/validators/is-short-iso-date-string");
Object.defineProperty(exports, "isShortISODateString", { enumerable: true, get: function () { return is_short_iso_date_string_1.isShortISODateString; } });
var is_valid_email_1 = require("./lib/validators/is-valid-email");
Object.defineProperty(exports, "isValidEmail", { enumerable: true, get: function () { return is_valid_email_1.isValidEmail; } });
var omit_1 = require("./lib/objects/omit");
Object.defineProperty(exports, "omit", { enumerable: true, get: function () { return omit_1.omit; } });
var generate_v4_uuid_1 = require("./lib/generators/generate-v4-uuid");
Object.defineProperty(exports, "generateV4UUID", { enumerable: true, get: function () { return generate_v4_uuid_1.generateV4UUID; } });
__exportStar(require("./lib/console/colors"), exports);
var normalizeHeaders_1 = require("./lib/converters/normalizeHeaders");
Object.defineProperty(exports, "normalizeHeaders", { enumerable: true, get: function () { return normalizeHeaders_1.normalizeHeaders; } });
var consts_1 = require("./lib/consts");
Object.defineProperty(exports, "ErrorMessages", { enumerable: true, get: function () { return consts_1.ErrorMessages; } });
var is_valid_http_url_1 = require("./lib/validators/is-valid-http-url");
Object.defineProperty(exports, "isValidHttpURL", { enumerable: true, get: function () { return is_valid_http_url_1.isValidHttpURL; } });
var is_valid_location_1 = require("./lib/validators/is-valid-location");
Object.defineProperty(exports, "isValidLocation", { enumerable: true, get: function () { return is_valid_location_1.isValidLocation; } });
var appendScriptWithAttributes_1 = require("./lib/browser/appendScriptWithAttributes");
Object.defineProperty(exports, "appendScriptWithAttributes", { enumerable: true, get: function () { return appendScriptWithAttributes_1.appendScriptWithAttributes; } });
