// @sitecore-cloudsdk/events/dist/cjs/src/lib/events/custom-event/form.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.form = void 0;
const internal_1 = require("@sitecore-cloudsdk/core/internal");
const utils_1 = require("@sitecore-cloudsdk/utils");
const initializer_1 = require("../../initializer/browser/initializer");
const sendEvent_1 = require("../send-event/sendEvent");
const custom_event_1 = require("./custom-event");
/**
 * A function that sends a form event to SitecoreCloud API
 *
 * @param formId - The required form ID string
 * @param interactionType - The required interaction type string. Possible values: "VIEWED", "SUBMITTED"
 * @param componentInstanceId - The required component instance ID string
 * @returns The response object that Sitecore EP returns or null
 */
async function form(formId, interactionType, componentInstanceId) {
    await (0, initializer_1.awaitInit)();
    const settings = (0, internal_1.getCloudSDKSettingsBrowser)();
    const id = (0, utils_1.getCookieValueClientSide)(settings.cookieSettings.name.browserId);
    const formEvent = new custom_event_1.CustomEvent({
        eventData: {
            extensionData: {
                componentInstanceId,
                formId,
                interactionType: interactionType.toUpperCase()
            },
            type: 'FORM'
        },
        id,
        sendEvent: sendEvent_1.sendEvent,
        settings
    });
    formEvent.page = undefined;
    return formEvent.send();
}
exports.form = form;
