// @sitecore-cloudsdk/personalize/dist/esm/src/lib/initializer/server/initializer.js
// © Sitecore Corporation A/S. All rights reserved. Sitecore® is a registered trademark of Sitecore Corporation A/S.
import { COOKIE_NAME_PREFIX, debug, enabledPackagesServer as enabledPackages, getCloudSDKRequest, getCloudSDKResponse, getCloudSDKSettingsServer, getEnabledPackageServer, PackageInitializerServer } from '@sitecore-cloudsdk/core/internal';
import { CloudSDKServerInitializer } from '@sitecore-cloudsdk/core/server';
import { ErrorMessages, PACKAGE_NAME, PERSONALIZE_NAMESPACE } from '../../consts';
import { createPersonalizeCookie } from './createPersonalizeCookie';
export async function sideEffects() {
    const cloudSDKSettings = getCloudSDKSettingsServer();
    const personalizeSettings = getEnabledPackageServer(PACKAGE_NAME)?.settings;
    debug(PERSONALIZE_NAMESPACE)('personalizeServer library initialized');
    if (!cloudSDKSettings.cookieSettings.enableServerCookie || !personalizeSettings.enablePersonalizeCookie)
        return;
    await createPersonalizeCookie(getCloudSDKRequest(), getCloudSDKResponse(), personalizeSettings, cloudSDKSettings);
}
/**
 * Makes the functionality of the personalize package available.
 *
 * @returns An instance of {@link CloudSDKServerInitializer}
 *
 */
export function addPersonalize(settings = { enablePersonalizeCookie: false }) {
    const cookieSettings = {
        name: {
            guestId: `${COOKIE_NAME_PREFIX}${getCloudSDKSettingsServer().sitecoreEdgeContextId}_personalize`
        }
    };
    const personalizeInitializer = new PackageInitializerServer({
        settings: { ...settings, cookieSettings },
        sideEffects
    });
    enabledPackages.set(PACKAGE_NAME, personalizeInitializer);
    return this;
}
CloudSDKServerInitializer.prototype.addPersonalize = addPersonalize;
export function verifyPersonalizePackageExistence() {
    if (!getEnabledPackageServer(PACKAGE_NAME))
        throw Error(ErrorMessages.IE_0017);
}
